### PR TITLE
[SWY-109] Add song to set action

### PR DESCRIPTION
**Pull request by justaddcl**
## Background
This PR is to add the "add song to set" feature where from the song details page, the user can quickly add the song to a set or create a set to add the song to.

## Summary by CodeRabbit

* **New Features**
  * Introduced a multi-step dialog for adding songs to sets, including set selection, section selection, song position, key selection, and review.
  * Added draggable and sortable song list components with drag-and-drop support.
  * Implemented flexible date picker supporting single, multiple, and range modes.
  * Added new UI components: Checkbox, MultiSelect, Progress bar, and enhanced Tooltip.
  * Provided event type selection with multi-select and improved filtering.
  * Added friendly date formatting for set and event displays.
  * Added detailed set and set section retrieval with authorization checks.
  * Added new API endpoints for upcoming and paginated sets with filtering and validation.

* **Enhancements**
  * Refined Card, HStack, and VStack components for better layout and styling flexibility.
  * Improved set and section selection flows with filtering and responsive design.
  * Updated UI icons for consistency and clarity.
  * Added support for tracking favorites and indexing in sets, songs, and event types.
  * Improved date picker labels, formatting, and UI controls.
  * Enhanced API with organization existence validation and improved error handling.
  * Improved form handling and callbacks in set creation and date picker fields.
  * Enhanced song content components with editable and read-only modes.
  * Improved responsive hooks with mobile detection.
  * Replaced static buttons with dialog components for better user interaction.

* **Bug Fixes**
  * Fixed module import path typo for HStack, ensuring correct component usage.
  * Added error handling and loading states in event type and set queries.
  * Corrected input handling and validation in song position and set section selection steps.

* **Chores**
  * Updated and added dependencies for drag-and-drop and UI libraries.
  * Added and reorganized exports for easier component imports.
  * Consolidated ESLint import sorting configuration.
  * Added VSCode settings to exclude specific packages from auto-import suggestions.
  * Added TypeScript declaration proxy for lucide-react module.
  * Reordered and cleaned import statements across multiple files.

* **Tests**
  * Added comprehensive tests for date formatting and number clamping utilities.

* **Documentation**
  * Improved inline comments and prop typing for clarity and maintainability.

### 1. Add Song to Set Dialog Trigger
**Location:** Song details page (`/[organization]/songs/[songId]`)
- **Test:** Click the "Add to a set" button in the song header
- **Expected:** Multi-step dialog opens showing "Select a set" as step 1/5

### 2. Step 1: Set Selection & Filtering
**Test filtering functionality:**
- **Event Type Filter:** Use the dropdown to filter by different event types
- **Date Range Filter:** Test the date picker with single dates, date ranges, and clearing filters
- **Combined Filters:** Apply both event type and date filters simultaneously
- **Filter Badges:** Verify dismissible filter badges appear and can be removed individually
- **Mobile vs Desktop:** Test filtering UI differences (drawer on mobile, horizontal bar on desktop)

**Test set selection:**
- **Upcoming Sets Section:** Verify "Next sets" shows chronologically ordered upcoming sets
- **Favorite Toggle:** Test the "See all" / "See only favorites" toggle functionality
- **All Sets Section:** Use "Load more sets" button to test infinite scrolling
- **Set Information:** Verify each set shows friendly date, event type, and song count
- **Tooltips:** Hover over sets within calendar week threshold to see full date tooltips

### 3. Step 2: Set Section Selection
**Test section management:**
- **Existing Sections:** Select from existing set sections and verify song lists display
- **Add New Section:** Click "Add section", select a section type, and create it
- **Section Validation:** Try creating duplicate section types (should prevent/handle gracefully)
- **Song Counts:** Verify section song counts are accurate in selection buttons

### 4. Step 3: Song Position Setting
**Test positioning functionality:**
- **Input Field:** Use the numeric input to set song position (test bounds: 1 to max+1)
- **Drag & Drop:** Drag the new song to different positions in the list
- **Visual Feedback:** Verify existing songs appear muted, new song appears highlighted
- **Position Sync:** Confirm input field updates when dragging, and vice versa

### 5. Step 4: Song Key Selection
**Test key selection:**
- **Key Grid:** Click different keys and verify selection highlighting
- **Preferred Key:** Verify the song's preferred key is highlighted with heart icon
- **Last Played Key:** Check if last played key appears with clock icon (if available)
- **Responsive Layout:** Test key grid layout on different screen sizes

### 6. Step 5: Review & Confirmation
**Test final review:**
- **Summary Display:** Verify all selections are correctly summarized (song, key, set, section)
- **Song List Preview:** Check the ordered song list shows the new song in correct position
- **Notes Field:** Add notes and verify they're included in the final submission
- **Submission:** Click "Add song to set" and verify success toast + data persistence

## Create New Set Flow
**Test set creation within dialog:**
- **Trigger:** Click "Create set" from the filtering section
- **Form Validation:** Test required fields (date, event type)
- **Success Flow:** Create set and verify it's automatically selected for song addition
- **Navigation:** After successful creation, verify the dialog continues to section selection

## Edge Cases & Error Handling
- **No Sets Available:** Test behavior when no sets match current filters
- **API Errors:** Test network failure scenarios (loading states, error messages)
- **Missing Data:** Test with songs that have no preferred key or play history
- **Navigation:** Test browser back/forward buttons during dialog flow
- **Dialog Closure:** Test closing dialog mid-flow and reopening (state reset)

## Mobile Responsiveness
**Test mobile-specific behaviors:**
- **Filter UI:** Verify filtering uses drawer interface on mobile
- **Dialog Layout:** Check dialog responsiveness and touch interactions
- **Drag & Drop:** Test touch-based drag and drop on mobile devices
- **Button Sizes:** Verify touch targets are appropriately sized

## Data Persistence Verification
**Test data integrity:**
- Navigate to the target set after adding a song to verify:
  - Song appears in correct position
  - Correct key is saved
  - Notes are preserved
  - Set song count updated correctly

### CODE REVIEW

**Comment by justaddcl** [comment]
#### 📝 Changed routes:


| [**…zation\]/songs/\[songId\]**](https://sanbi-18pqcx5rt-justaddcls-projects.vercel.app/[organization]/songs/[songId]) | [**/**](https://sanbi-18pqcx5rt-justaddcls-projects.vercel.app/) |
|:-:|:-:|
|<a href="https://sanbi-18pqcx5rt-justaddcls-projects.vercel.app/[organization]/songs/[songId]">Screenshot of /[organization]/songs/[songId]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/2RuBzTAR3CjcQ6FpzsH2cAe_vClpkx-WwNepitA83z4VAoLZGCeGbwxoKGvsVjVmBBFX8HV46fYTCtxmC2Yafg.png">(view full size)</a> |<a href="https://sanbi-18pqcx5rt-justaddcls-projects.vercel.app/">Screenshot of /</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/1tyIq9g7m-e6QSAlB7XXv2ykoeMN6GatyBrJaSwUBuKsulXK0tJxE8YLku5QqIUIsaDdEBuxN7DRAyW9QXvL2Q.png">(view full size)</a> |







Commit 45a9e74036e59889b2c84e9345312b4b5b82b220 (https://sanbi-18pqcx5rt-justaddcls-projects.vercel.app).

File: src/app/layout.tsx [hunk 1]
```diff
@@ -39,7 +39,7 @@ export default function RootLayout({
         <head>
           <meta
             name="viewport"
-            content="width=device-width, initial-scale=1, maximum-scale=1"
+            content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no"
           />
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/app/layout.tsx]: @coderabbitai I added this change in an attempt to stop Safari and Chrome on iOS to automatically zoom into text inputs when selected on mobile. Other web apps I use prevent this default behavior. Is there a better way to accomplish this?

File: src/modules/songs/forms/AddSongToSet/components/SetSelectionStep.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+          <SetSelectionSetItem
+            title={"Sun, May 11"}
+            subtitle={"Sunday Service"}
+            label={" 3 songs"}
+          />
+          <SetSelectionSetItem
+            title={"Sun, May 18"}
+            subtitle={"Sunday Service"}
+            label={"4 songs"}
+          />
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/AddSongToSet/components/SetSelectionStep.tsx]: @coderabbitai This was just placeholder content so should not be an issue with my changes since.

File: src/modules/shared/components/DraggableSongList/DraggableSongList.tsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+              <SongContent
+                songKey={activeSongItem.songKey}
+                name={activeSongItem.name}
+                index={activeSongItem.index}
+              />
+            </HStack>
+          ) : null}
+        </DragOverlay>,
+        document.body,
+      )}
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/shared/components/DraggableSongList/DraggableSongList.tsx:153]: @coderabbitai Question for clarification, can a user who is using a screen reader use the drag and drop feature? I do want this to be keyboard accessible, but not sure how the role and aria-label will help with that. Can you help me better understand that?

[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/shared/components/DraggableSongList/DraggableSongList.tsx:153]: @coderabbitai Thank you for the explanation. I'm still in the proof of concept stage, but please remember this item for when I'm focusing on improving the quality of the product when getting ready to launch!

File: src/modules/songs/forms/AddSongToSet/components/ReviewStep/ReviewStep.tsx [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        async onSuccess() {
+          toast.success("Song added to set!", { id: toastId });
+
+          await apiUtils.song.get.invalidate({ songId: song.id });
+          await apiUtils.set.get.invalidate({ setId: selectedSetId });
+          // FIXME: doesn't seem to update the SSR data on the song page
+          // await apiUtils.song.getPlayHistory.invalidate({
+          //   songId: song.id,
+          //   // organizationId: userMembership.organizationId,
+          // });
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/modules/songs/forms/AddSongToSet/components/ReviewStep/ReviewStep.tsx:110]: @coderabbitai Yes, I would like you to help me investigate what the most effective way to update the SSR data may be. (I suspect it may be a full refresh, but would love there to be another way)

**Comment by justaddcl** [comment]
@coderabbitai How would you suggest I test the changes in this PR? Please make sure that the instructions are appropriately thorough, but are only kept to the scope of the changes in this PR and where it can be tested by looking or interacting with the UI. Please also take a look at previous PRs to make sure we're not re-testing something already covered by a different PR. The tests should only cover what was added here.

**Comment by justaddcl** [comment]
@coderabbitai This is a very large PR that has been open for a long time so I'm keen to merge and deploy if possible. Since I am still in the proof of concept stage, I'm not yet addressing all of the nit picks that would add refinement and polish. What are the critical changes still needed on this PR (if any) and what are the nit-picky change that would enhance the experience? I want to clearly prioritize all these items to know what needs to be done to close the PR.
